docs/style-guide.md

```diff
@@ -0,0 +1,439 @@
+
+```
+fn function_name(param: ParamType) -> ReturnType { … }
+```
+
+may still have parameters "passed by reference" when it can determine that the reference is no longer subsequently used by the calling code (this is a "move" in Rust parlance).
+
+Furthermore, it will often [ellide the copy[(https://en.wikipedia.org/wiki/Copy_elision) of the return value if it determines the returned object would otherwise go out of scope. It will allocate the space for the return value on the caller's stack or use the memory in a destination struct, to directly write the "returned" value in its final destination with no copying whatsoever. This is called Return Value Optimization (RVO) and while it is now available in C++ as well, it's a relatively new feature there.
+
+It is still often better (for reasons of borrowing and ownership) to pass structs by non-mutable reference, but returning **newly created** results by value (even potentially large structures) is not expected to cause performance issues.
+
+### Return new instance by value :: suggested
+
+Use "return by value" for any functions which produce new data and let Rust handle optimizing away any copying, unless there's really measurable performance issue.
+
+### Pass struct parameters by reference where possible :: suggested
+
+When passing struct types to functions, it's always semantically safe to pass a reference (since there can be no race conditions around its use and it cannot be modified unexpectedly while being processed). The called function should be responsible for taking a copy if it needs to (e.g. by cloning or via the ToOwned trait).
+
+### Pass Option<T> by value where possible :: suggested
+
+[Option](https://doc.rust-lang.org/std/option/enum.Option.html) is a bit special, and Rust goes to great lengths to avoid needing to allocate an additional instance for this type. In particular, `Option<char>` is a zero-allocation representation and should be passed by value. This applies to some types where the extra [None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None) state can be encoded as an otherwise "invalid" bit pattern.
+
+In the case of `Option<char>`, `None` is [encoded as `0x110000`](https://rust.godbolt.org/z/-ZFwKB) or other invalid bit patterns. This means types like `Vec<Option<char>>` or `[Option<char>;N]` should be preferred over `Vec<&Option<char>>` or `[&Option<char>;N]` which will often remove the need for additional lifetimes.
+
+There are times where this may not be possible (e.g. if using non-`Copy` types you may need `&Option<T>` or `Option<&T>`), and there's a trade off between designing `Copy` types, exppected to be smaller value types, and other more complex types.
+
+### Pass Box<T> by value where possible :: suggested
+
+It's unlikely we will have many `Box<T>`s around. We should return boxes by-value, and accept things as `&T` as much as possible, instead of `&Box<T>`.
```

Comment:
The exception here are arrays/slices. In `Locale` I do use `Box<[T]>` over `Vec<T>` since it guarantees layout and immutability and is generally cheaper.
For scenarios where we'll want to hold immutable list (as in, if you want to change the list, you assign a new value, rather than modify the list), `Box<[T]>` is a good choice.

But it should be an implementation detail and not part of a public API - so `Deref` to `&[T]`.